### PR TITLE
Dispatch textDocument/didChange after rename

### LINKED ISSUE #3343: Send didChange/didSave events after rename
Rename operations modify files other than the current buffer, and language servers (and `tsserver`) need to know that these files are changed. We should make ALE tell language servers that files are changed or saved when they change as the result of a code action or rename operation. This will keep type information up to date when you rename something across many files from outside of the file where a name originates.

### PR DESCRIPTION
**Pull request by jeremija**
Previously whenever we renamed a symbol that was referenced from other files we'd just edit those files in the background, and the LSP wouldn't know about these changes. If we tried to rename the same symbol again, the renaming would fail. In some scenarios, the operation would just be wrong. Here is an attempt to fix this issue.

I also noticed another bug when using Go with `gopls` LSP and the `gofmt` fixer. Whenever the file was saved, the `gofmt` would run and reformat the file. But it seems there was some kind of a race condition so I disabled saving for now, and all of the modified files will be unsaved, so the user should call `:wa` to save them. I personally like this even better because I can inspect exactly what changes happened, and I instantly see them in the other opened buffers, which was previously not the case.

Finally, this change fixes a bug when trying to replace the contents of a hidden buffer.

Fixes #3343, #3642, #3781.

Video example:

Rename example

### CODE REVIEW

File: test/test_code_action_corner_cases.vader [hunk 1]
```diff
@@ -92,7 +99,7 @@ Execute(Preserve (no)eol at eof):
 
   " there doesn't seem to be a way to tell if a buffer is empty or contains one
   " empty line :-(
-  AssertEqual "", TestChanges("", [], 'file')
+  AssertEqual "\n", TestChanges("", [], 'file')
```
Comment (liskin) [inline test/test_code_action_corner_cases.vader:102]: This makes the above comment meaningless.

The point of that comment was that this test can't be done for the `save` and `buffer` modes, as these go through a buffer and change the file contents even though they should not. Now that all edits go through a buffer, not even the `file` mode works properly. At this point we might as well test all three modes and add a comment explaining that the behaviour is wrong and the only point of the test will be to catch changes in behaviour, not to test that it behaves correctly.

File: test/test_code_action_corner_cases.vader [hunk 2]
```diff
@@ -28,6 +28,11 @@
 "   $ node test_code_action_corner_cases.js <test_code_action_corner_cases.vader
 
 Before:
+  runtime autoload/ale/lsp.vim
+
+  function! ale#lsp#NotifyForChanges(conn_id, buffer) abort
```
Comment (liskin) [inline test/test_code_action_corner_cases.vader:33]: Tests that override ale functions should `runtime` the overridden files again in `After:`

**Comment by liskin** [comment]
I don't really know enough about vim scripting to be able to tell whether doing `:edit` and `:w!` is a good idea, so I only commented on the parts that I do understand. But overall it looks like it might just work. :-)

File: test/test_code_action_corner_cases.vader (hunk 1 above)
Comment (liskin) [reply test/test_code_action_corner_cases.vader:102]: Also, since all modes go through a buffer, it may not be necessary to test everything three times. The only differences now are whether the buffer is switched and whether it's saved (the test plumbing would save it anyway if it was meant to not be saved), none of which is relevant to these corner case tests any more.

Comment (jeremija) [reply test/test_code_action_corner_cases.vader:102]: Thanks, I was going to ask you if it's okay to remove this line!

File: test/test_code_action_corner_cases.vader (hunk 2 above)
Comment (jeremija) [reply test/test_code_action_corner_cases.vader:33]: Yeah that's way better than what I did, will update it!

Comment (jeremija) [reply test/test_code_action_corner_cases.vader]: Done!

File: test/test_code_action_corner_cases.vader (hunk 1 above)
Comment (jeremija) [reply test/test_code_action_corner_cases.vader]: Done!

**Comment by daliusd** [comment]
So it is closed. No need to review anymore. Anyway @liskin covered everything perfectly.

**Comment by jeremija** [comment]
Oops if I closed this, it must've been by accident. Reopening now, sorry about this!

**Comment by daliusd** [comment]
Overall I usually forget about `ALERename` and when I do I follow up it with `:wa`. So overall I think this is good fix.

**Comment by bogdan** [comment]
@jeremija I tried your branch with tsserver and tried to rename `A` to `A1` from `b.ts` and here is what I get:

image

**Comment by liskin** [comment]
I kind of forgot to review the followups to my previous review, so if this ever gets reopened and the issue reported by @bogdan is dealt with, I'll be happy to publish some fixups to better address my concerns about the tests (I vaguely remember I wanted to do more than the current version of this PR does).

Oh and unfortunately I still haven't managed to actually test this fix because I don't use any language/LSP combination where I'd need code actions across multiple files. :-/

**Comment by jeremija** [comment]
Hey @liskin, @bogdan sorry for going MIA.

@bogdan I'm failing to reproduce your error, and I've been using this fork since I opened this MR full time.

Any chance you can post your `ALEInfo` and vim/nvim version?

I'd love to have this merged because I really think it improves the usability.

**Comment by hsanson** [comment]
@jeremija cannot reopen the MR, Github tells me is because the branch has been re-created or force-pushed.

**Comment by jeremija** [comment]
@hsanson shall I open a new one instead?

**Comment by liskin** [comment]
> Hey @liskin, @bogdan sorry for going MIA.

Happens to all of us :-)

(I've had a taskwarrior task to review this PR since 2021-07-14 which I kept replanning to "tomorrow" until 2021-08-29, when I decided to stop lying to myself and just unscheduled it :-))

**Comment by jeremija** [comment]
Ah sorry about that @liskin 🙂
